File: Python/704_Binary_Search.py
"""
Given an array of integers nums which is sorted in ascending order, and an integer target, write a function to search target in nums. If target exists, then return its index. Otherwise, return -1.

You must write an algorithm with O(log n) runtime complexity.
"""
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Define target in the range of left-closed and right-closed, [left, right]
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] > target:
                right = mid - 1 # target is in the left interval, so [left, middle - 1]
            elif nums[mid] == target:
                return mid
            else:
                left = mid + 1 # target is in the right interval, so [middle + 1, right]
        # target value not found
        return -1                 

File: Java/704_Binary_Search.java
/**
 * Since the target is defined within the [left, right] interval, there are two points to note:

Use while (left <= right) because left == right is meaningful, so <= should be used.
If nums[middle] > target, right should be assigned the value middle - 1 because the current nums[middle] is definitely not the target. Therefore, the ending index of the left interval to be searched next is middle - 1.
 * 
 */
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right-left) / 2;
            if (nums[mid] > target) {
                right = mid - 1;
            } else if (nums[mid] < target) {
                left = mid + 1;
            }else {
                return mid;
            }
        }
        return -1;
    }
}